File: selenium-csharp-e2e-template/Pages/LoginPage.cs
using OpenQA.Selenium;

namespace Pages;

public class LoginPage : BasePage
{
    private static readonly By UsernameInput = By.CssSelector("input[name='username'], input#username, input[type='email']");
    private static readonly By PasswordInput = By.CssSelector("input[name='password'], input#password");
    private static readonly By SubmitButton = By.CssSelector("button[type='submit'], input[type='submit']");

    public LoginPage(IWebDriver driver) : base(driver) { }

    public void Login(string username, string password)
    {
        WaitForElement(UsernameInput).SendKeys(username);
        Driver.FindElement(PasswordInput).SendKeys(password);
        Driver.FindElement(SubmitButton).Click();
    }
}

File: web/selenium-typescript/pages/loginPage.ts
import { By } from 'selenium-webdriver';
import { BasePage } from './basePage';

export class LoginPage extends BasePage {
  private readonly usernameInput = By.css('input[name="username"], input#username, input[type="email"]');
  private readonly passwordInput = By.css('input[name="password"], input#password');
  private readonly submitButton = By.css('button[type="submit"], input[type="submit"]');

  async login(username: string, password: string): Promise<void> {
    const usernameElement = await this.waitForElement(this.usernameInput);
    await usernameElement.sendKeys(username);
    
    const passwordElement = await this.driver.findElement(this.passwordInput);
    await passwordElement.sendKeys(password);
    
    const submitElement = await this.driver.findElement(this.submitButton);
    await submitElement.click();
  }
}